File: leetcode-practice-app/src/data/64. Minimum Path Sum.cpp
const int inf = 0x3f3f3f3f;
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int H = grid.size(),
            W = grid[0].size();
        
        vector<vector<int>> dp(H, vector<int>(W, inf));
        for(int row = 0; row < H; row++){
            for(int col = 0; col < W; col++) {
                if(row == 0 && col == 0) {
                    dp[row][col] = grid[row][col];
                }else {
                    dp[row][col] = grid[row][col] + min(( (row == 0) ? inf : dp[row-1][col]), ((col == 0) ? inf : dp[row][col-1]));   
                }
            }
        }
        
        return dp[H-1][W-1];
    }
};

File: solutions/64. Minimum Path Sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        if (grid == null)
            return 0;

        int R = grid.length;
        int C = grid[0].length;

        for (int col = 1; col < C; col++) {
            grid[0][col] += grid[0][col - 1];
        }

        for (int row = 1; row < R; row++) {
            grid[row][0] += grid[row - 1][0];
        }

        for (int row = 1; row < R; row++) {
            for (int col = 1; col < C; col++) {
                if (row - 1 < 0 || col - 1 < 0)
                    continue;
                int cost = getCost(grid, row, col);
                grid[row][col] = cost;
            }
        }

        return grid[R - 1][C - 1];
    }

    private int getCost(int[][] grid, int row, int col) {
        return Math.min(grid[row - 1][col], grid[row][col - 1]) + grid[row][col];
    }
}
